### PR TITLE
improvements in translations, casting home rules

### PR DESCRIPTION
**Pull request by silentmark**
HI @moo-man , 

I would like to add few things to your system: 
1. Some improvements for translations - i've normalized some of the arrays in Config and moved English versions into language file. I've also moved initialization of objects that would be harder to automatically localize into dedicated method inside Init hook, 
2. I've added some homerules that makes spell casting and channeling even more deadly and crazy. Those rules were inspired with Unofficial Grimoire. i would move them to separate module, if it wouldn't require updates in the channeling/casting test dialog. 
3. I've added one handlebar and small change in test dialog for customized fumbles, that i want to added in separate module. 

Last point - would it be fine to include Polish language directly in core system (only the pl.json file in lang folder + entry in system.json) or do you prefer not to pollute your repo and keep it as separate module? Currently i haven't add it to PR.

### CODE REVIEW

File: lang/en.json [hunk 1]
```diff
@@ -102,6 +101,8 @@
...
     "Configure": "Configure",
```
Comment (moo-man) [inline lang/en.json:106]: Can this be a link to the information (such as the wiki or the pdf itself) instead of a text dump in the settings?

Comment (silentmark) [reply lang/en.json:106]: Sure, didn't think about it. I will prepare something for wiki and find the link to pdf itself

Comment (silentmark) [reply lang/en.json:106]: I've made a copy of your wiki and added suggested changes here: https://github.com/silentmark/wfrp4e-wiki/blob/master/Homebrew-Rules.md#magic-insiperd-by-unofficial-grimoire

I think there is no way to create a PR to github wiki

Comment (moo-man) [reply lang/en.json]: The wiki should be publicly editable, is it not?

Comment (silentmark) [reply lang/en.json]: If it is, I couldn't find a way to modify it directly

Comment (silentmark) [reply lang/en.json]: Anyway, let me know if there could be any improvements or changes on my side

File: modules/actor/actor-wfrp4e.js [hunk 2]
```diff
@@ -894,6 +894,16 @@ export default class ActorWfrp4e extends Actor {
...
 
+    //@HOUSE
+    testData.unofficialGrimoire = game.settings.get("wfrp4e", "unofficialgrimoire");
+    let advantages = this.status.advantage.value || 0;
+    if (testData.unofficialGrimoire) {
+      advantages = "N/A";
+    } else {
+      this.status.advantage.value || 0
+    }
+    //@HOUSE
```
Comment (moo-man) [inline modules/actor/actor-wfrp4e.js:905]: Can you add  `game.wfrp4e.utility.logHomebrew("unofficialgrimoire")` inside the true block?

File: modules/actor/actor-wfrp4e.js [hunk 3]
```diff
@@ -918,6 +929,11 @@ export default class ActorWfrp4e extends Actor {
         testData.testDifficulty = game.wfrp4e.config.difficultyModifiers[html.find('[name="testDifficulty"]').val()];
         testData.successBonus = Number(html.find('[name="successBonus"]').val());
         testData.slBonus = Number(html.find('[name="slBonus"]').val());
+        if (testData.unofficialGrimoire) {
+          testData.ingredientMode = html.find('[name="ingredientTypeSelected"]').val();
+          testData.overchannelling = Number(html.find('[name="overchannelling"]').val());
+          testData.quickcasting = html.find('[name="quickcasting"]').is(':checked');
+        }
```
Comment (moo-man) [inline modules/actor/actor-wfrp4e.js:936]: Can you add  `game.wfrp4e.utility.logHomebrew("unofficialgrimoire")`

File: modules/actor/actor-wfrp4e.js [hunk 4]
```diff
@@ -1021,6 +1039,9 @@ export default class ActorWfrp4e extends Actor {
         testData.testDifficulty = game.wfrp4e.config.difficultyModifiers[html.find('[name="testDifficulty"]').val()];
         testData.successBonus = Number(html.find('[name="successBonus"]').val());
         testData.slBonus = Number(html.find('[name="slBonus"]').val());
+        if (testData.unofficialGrimoire) {
+          testData.ingredientMode = html.find('[name="ingredientTypeSelected"]').val();
+        }
```
Comment (moo-man) [inline modules/actor/actor-wfrp4e.js:1044]: Can you add  `game.wfrp4e.utility.logHomebrew("unofficialgrimoire")`

File: modules/hooks/init.js [hunk 5]
```diff
@@ -566,6 +566,29 @@ export default function() {
...
+    // Register Partial Channelling
+    game.settings.register("wfrp4e", "partialChannelling", {
+      name: "SETTINGS.PartialChannelling",
+      hint: "SETTINGS.PartialChannellingHint",
+      scope: "world",
+      homebrew: true,
+      config: false,
+      default: false,
+      type: Boolean
+    });
```
Comment (moo-man) [inline modules/hooks/init.js:578]: Why is this setting duplicated?

File: modules/system/rolls/cast-test.js [hunk 6]
```diff
@@ -9,6 +9,10 @@ export default class CastTest extends TestWFRP {
 
     this.preData.itemData = data.itemData || this.item.toObject() // Store item data to avoid rerolls being affected by changed channeled SL
     this.preData.skillSelected = data.skillSelected;
+    this.preData.overchannelling = data.overchannelling;
+    this.preData.quickcasting = data.quickcasting;
+    this.preData.unofficialGrimoire = data.unofficialGrimoire;
+    this.preData.ingredientMode = data.ingredientMode;
```
Comment (moo-man) [inline modules/system/rolls/cast-test.js:15]: Isn't this all unofficial grimoire? so an unofficialGrimoire object that holds all these flags should work?

File: modules/system/rolls/cast-test.js [hunk 7]
```diff
@@ -44,6 +48,11 @@ export default class CastTest extends TestWFRP {
   runPreEffects() {
     super.runPreEffects();
     this.actor.runEffects("preRollCastTest", { test: this, cardOptions: this.context.cardOptions })
+    //@HOUSE
+    if (this.preData.unofficialGrimoire && this.preData.ingredientMode == 'power' && this.hasIngredient) {
+      this.preData.canReverse = true;
+    }
+    //@HOUSE
```
Comment (moo-man) [inline modules/system/rolls/cast-test.js:55]: Can you add  `game.wfrp4e.utility.logHomebrew("unofficialgrimoire")`

File: modules/system/rolls/cast-test.js [hunk 8]
```diff
@@ -105,15 +123,27 @@ export default class CastTest extends TestWFRP {
         }
         //@/HOUSE
       }
+      //@/HOUSE
+      if (this.preData.overchannelling > 0) { 
+        this.result.tooltips.miscast.push(game.i18n.localize("CHAT.OverchannellingMiscast"))
+        miscastCounter++;
+      }
+      //@/HOUSE
```
Comment (moo-man) [inline modules/system/rolls/cast-test.js:131]: Can you add  `game.wfrp4e.utility.logHomebrew("overchannelling")`

File: modules/system/rolls/cast-test.js [hunk 9]
```diff
@@ -105,15 +123,27 @@ export default class CastTest extends TestWFRP {
...
     {
       this.result.castOutcome = "failure"
       this.result.description = game.i18n.localize("ROLL.CastingFailed")
-
+      //@/HOUSE
+      if (this.preData.overchannelling > 0) { 
+        this.result.tooltips.miscast.push(game.i18n.localize("CHAT.OverchannellingMiscast"))
+        miscastCounter++;
+      }
+      //@/HOUSE
```
Comment (moo-man) [inline modules/system/rolls/cast-test.js:142]: Can you add  `game.wfrp4e.utility.logHomebrew("overchannelling")`

File: modules/system/rolls/cast-test.js [hunk 10]
```diff
@@ -124,7 +154,13 @@ export default class CastTest extends TestWFRP {
     else // Successful test, casted - determine overcast
     {
       this.result.castOutcome = "success"
-      this.result.description = game.i18n.localize("ROLL.CastingSuccess")
+      this.result.description = game.i18n.localize("ROLL.CastingSuccess");
+      //@/HOUSE
+      if (this.preData.overchannelling > 0) {
+        slOver += this.preData.overchannelling;
+      }
+      //@/HOUSE
```
Comment (moo-man) [inline modules/system/rolls/cast-test.js:162]: Can you add  `game.wfrp4e.utility.logHomebrew("overchannelling")`

File: modules/system/rolls/cast-test.js [hunk 11]
```diff
@@ -141,8 +177,13 @@ export default class CastTest extends TestWFRP {
...
       }
       //@/HOUSE
-
     }
+    //@HOUSE
+    if (this.preData.quickcasting && miscastCounter > 0) { 
+      this.result.other.push(game.i18n.localize("CHAT.Quickcasting"))
+      miscastCounter++;
+    }
+    //@/HOUSE
```
Comment (moo-man) [inline modules/system/rolls/cast-test.js:186]: Can you add  `game.wfrp4e.utility.logHomebrew("quickcasting")`

File: modules/system/rolls/cast-test.js [hunk 12]
```diff
@@ -182,9 +223,19 @@ export default class CastTest extends TestWFRP {
...
+      if (this.hasIngredient && this.item.ingredient.quantity.value > 0 && !this.context.edited && !this.context.reroll) {
+        this.item.ingredient.update({ "data.quantity.value": this.item.ingredient.quantity.value - 1 })
+        ChatMessage.create({ speaker: this.data.context.speaker, content: game.i18n.localize("ConsumedIngredient") })
+      }
+    //@/HOUSE
+    } else {
+      // Find ingredient being used, if any
+      if (this.hasIngredient && this.item.ingredient.quantity.value > 0 && !this.context.edited && !this.context.reroll)
+        this.item.ingredient.update({ "data.quantity.value": this.item.ingredient.quantity.value - 1 })
+    }
```
Comment (moo-man) [inline modules/system/rolls/cast-test.js:238]: Can you add... you guessed it...   `game.wfrp4e.utility.logHomebrew("unofficialgrimoire")`

File: templates/chat/roll/test-card.html [hunk 13]
```diff
@@ -54,7 +54,29 @@
...
+                {{#if (tableExists "oopsblades")}}
+                <li><a class ="table-click fumble-roll" data-table = "oopsblades"><i class="fas fa-list"></i>{{test.result.fumble}} (Sieczna)</a></li>
+                {{/if}}
+                {{#if (tableExists "oopspummelling")}}
+                <li><a class ="table-click fumble-roll" data-table = "oopspummelling"><i class="fas fa-list"></i>{{test.result.fumble}} (Obuchowa)</a></li>
+                {{/if}}
+                {{#if (tableExists "oopsbrawling")}}
+                <li><a class ="table-click fumble-roll" data-table = "oopsbrawling"><i class="fas fa-list"></i>{{test.result.fumble}} (Bijatyka)</a></li>
+                {{/if}}
+            </ul>
```
Comment (moo-man) [inline templates/chat/roll/test-card.html:79]: I'm not sure I like this (and there's still some Polish here)

The way I've seen people do this before is use multi-columned oops tables, so when you click on the table links, it lists the columns to roll on, instead of crowding the test card with all these links.

File: modules/actor/actor-wfrp4e.js (hunk 2 above)
Comment (silentmark) [reply modules/actor/actor-wfrp4e.js]: done

File: modules/actor/actor-wfrp4e.js (hunk 3 above)
Comment (silentmark) [reply modules/actor/actor-wfrp4e.js]: done

File: modules/actor/actor-wfrp4e.js (hunk 4 above)
Comment (silentmark) [reply modules/actor/actor-wfrp4e.js]: done

File: modules/hooks/init.js (hunk 5 above)
Comment (silentmark) [reply modules/hooks/init.js]: this mistake has been sponsored by our one and only duet Copy & Paste. Fixed

File: modules/system/rolls/cast-test.js (hunk 6 above)
Comment (silentmark) [reply modules/system/rolls/cast-test.js]: done, changed into object

File: modules/system/rolls/cast-test.js (hunk 7 above)
Comment (silentmark) [reply modules/system/rolls/cast-test.js]: done

File: modules/system/rolls/cast-test.js (hunk 8 above)
Comment (silentmark) [reply modules/system/rolls/cast-test.js]: done

File: modules/system/rolls/cast-test.js (hunk 9 above)
Comment (silentmark) [reply modules/system/rolls/cast-test.js]: done

File: modules/system/rolls/cast-test.js (hunk 10 above)
Comment (silentmark) [reply modules/system/rolls/cast-test.js]: done

File: modules/system/rolls/cast-test.js (hunk 11 above)
Comment (silentmark) [reply modules/system/rolls/cast-test.js]: done

File: modules/system/rolls/cast-test.js (hunk 12 above)
Comment (silentmark) [reply modules/system/rolls/cast-test.js]: done

File: templates/chat/roll/test-card.html (hunk 13 above)
Comment (silentmark) [reply templates/chat/roll/test-card.html]: ok, i checked the tableroll method and play around with it, for now it was removed, thanks

**Comment by silentmark** [comment]
i've made the updates, but i havent checked all of them yet, will make some tests later this week and push fixes in case of any issues

File: modules/system/config-wfrp4e.js [hunk 14]
```diff
@@ -1394,232 +1612,17 @@ WFRP4E.effectTextStyle = CONFIG.canvasTextStyle.clone();
...
+    "equipped" : "WFRP4E.effectApplication.equipped",
+    "apply" : "WFRP4E.effectApplication.apply",
+    "damage" : "WFRP4E.effectApplication.damage",
 }
 
 WFRP4E.applyScope = {
-    "actor" : "Actor",
-    "item" : "Item"
+    "actor" : "WFRP4E.applyScope.Aktor",
+    "item" : "WFRP4E.applyScope.Przedmiot"
```
[the remark this answers is not in the document]
Comment (silentmark) [reply modules/system/config-wfrp4e.js]: thanks, fixed

File: modules/system/config-wfrp4e.js [hunk 15]
```diff
@@ -782,528 +784,744 @@ WFRP4E.actorSizeEncumbrance = {
...
-                        "effectData": {
-                            "description": "Tests to affect",
-                            "slBonus": "-1"
+        stomp : {
+            name: game.i18n.localize("NAME.Stomp"),
+            type: "trait",
+            effects : [],
+            data: {
+                specification: { value: "4" },
+                rollable: { value: true, rollCharacteristic: "ws", bonusCharacteristic: "s", defaultDifficulty: "challenging", damage : true, skill : "Broń Biała (Bijatyka)" },
```
[the remark this answers is not in the document]
Comment (silentmark) [reply modules/system/config-wfrp4e.js]: thanks, fixed

**Comment by silentmark** [comment]
OK, so i made some tests and it looks to work well, found one issue with success calculation for channeling (some obsolete code) and one problem with translations related to condition application homebrew which was relying on object that was not yet present.

File: lang/en.json [hunk 16]
```diff
@@ -7,7 +7,6 @@
     "ACTOR.TypeCreature" : "Creature",
     "ACTOR.TypeVehicle" : "Vehicle",
 
-    "BROWSER.Button": "Browser",
```
Comment (moo-man) [inline lang/en.json:10]: Why was this removed?

File: modules/system/config-wfrp4e.js [hunk 17]
```diff
@@ -782,528 +784,744 @@ WFRP4E.actorSizeEncumbrance = {
     "mnst": 100
 }
 
+WFRP4E.PrepareSystemItems = function() {
```
Comment (moo-man) [inline modules/system/config-wfrp4e.js:787]: This was added to support localization? Other reasons?

File: lang/en.json (hunk 16 above)
Comment (silentmark) [reply lang/en.json:10]: There was a duplicate.

File: modules/system/rolls/channel-test.js [hunk 18]
```diff
@@ -149,11 +171,26 @@ export default class ChannelTest extends TestWFRP {
...
+      game.wfrp4e.utility.logHomebrew("unofficialgrimoire-overchannelling");
+      this.result.overchannelling = this.item.cn.SL + SL - this.item.cn.value;
+    }
+    //@HOUSE
+  
+    if(SL > 0) {
+      this.result.SL = "+" + SL;
+    } else {
+      this.result.SL = SL.toString()
+    }
```
Comment (moo-man) [inline modules/system/rolls/channel-test.js:186]: This is outside the homebrew sections, is it a fix for something?

File: templates/chat/roll/channel-card.html [hunk 19]
```diff
@@ -1,4 +1,12 @@
...
-</div>
 
+  {{#if test.result.overchannelling}}
+    <div class="card-content">
+    {{localize "Overchannelling"}}: <span class="overchannelling-count"><b>{{test.result.overchannelling}}</b></span> 
+    </div>
+  {{/if}}
+    <div class="card-content">
+      {{localize "SHEET.CN"}}: <span class="spellcn-count"><b>{{test.result.CN}}</b></span> 
+    </div>
```
Comment (moo-man) [inline templates/chat/roll/channel-card.html:11]: Is there a good reason to show the CN? Personally as a GM I wouldn't want that shown.

File: modules/apps/stat-parser.js [hunk 20]
```diff
@@ -188,7 +188,7 @@ export default class StatBlockParser extends FormApplication {
             catch {}
             if (!skillItem) {
                 console.error("Could not find " + skill)
-                ui.notifications.error(game.i18n.format("ERRORParser", {name: skill}), { permanent: true })
+                ui.notifications.error(game.i18n.format("ERROR.Parser", {name: skill}), { permanent: true })
```
Comment (moo-man) [inline modules/apps/stat-parser.js:191]: I've updated the stat parser so pull from develop, the conflicts should be simple to resolve.

File: templates/chat/roll/channel-card.html (hunk 19 above)
Comment (silentmark) [reply templates/chat/roll/channel-card.html:11]: For me it was useful, to know how much channelling left, i can remove this though or make it simply for gm-only.

File: modules/system/rolls/channel-test.js (hunk 18 above)
Comment (silentmark) [reply modules/system/rolls/channel-test.js:186]: yes, as some other aspects are modifying the SLs, but never update the value in the result object, and this leads to different values shown on card vs actually applied, i.e. this part: 

      if (Number(SL) == 0 && game.settings.get("wfrp4e", "extendedTests"))
          SL = 1;

**Comment by silentmark** [comment]
@moo-man , another bunch of comments from my side.

File: templates/chat/roll/channel-card.html (hunk 19 above)
Comment (silentmark) [reply templates/chat/roll/channel-card.html:12]: added if user is GM.

File: templates/chat/roll/channel-card.html [hunk 21]
```diff
@@ -6,7 +6,9 @@
     {{localize "Overchannelling"}}: <span class="overchannelling-count"><b>{{test.result.overchannelling}}</b></span> 
     </div>
   {{/if}}
+  {{#if user.isGM}}
     <div class="card-content">
       {{localize "SHEET.CN"}}: <span class="spellcn-count"><b>{{test.result.CN}}</b></span> 
     </div>
+  {{/if}}
 </div>
```
[the remark this answers is not in the document]
Comment (moo-man) [reply templates/chat/roll/channel-card.html:14]: Now that I'm at my computer: 

Notice the `hide-option` class
https://github.com/moo-man/WFRP4e-FoundryVTT/blob/a5d52c49cc54a215b61e306259de6345d63bad0d/templates/chat/roll/test-card.html#L23

Removed in the `renderChatMessage` hook
https://github.com/moo-man/WFRP4e-FoundryVTT/blob/a5d52c49cc54a215b61e306259de6345d63bad0d/modules/hooks/chat.js#L285

Note that the code is *very* old and I haven't looked at it in a long time. I would presume it would still work by just adding `.hide-option` to the element.

[the remark this answers is not in the document]
Comment (silentmark) [reply templates/chat/roll/channel-card.html:12]: Ok, done this way, thanks.

[the remark this answers is not in the document]
Comment (silentmark) [reply templates/chat/roll/channel-card.html:12]: Ok, i noticed your second comment a bit later, but i think my dedicated class and dedicated removal will be better, as i dont want to interfere and depend on the roll-test setting
